### PR TITLE
Add MaxRestarts to RestartWithBackoff

### PR DESCRIPTION
**Pull request by ismaelhamed**
Port [#24328](https://github.com/akka/akka/pull/24328)

### CODE REVIEW

File: src/core/Akka.API.Tests/CoreAPISpec.ApproveStreams.approved.txt [hunk 1]
```diff
@@ -1528,17 +1528,21 @@ namespace Akka.Streams.Dsl
     }
     public class static RestartFlow
     {
-        public static Akka.Streams.Dsl.Flow<TIn, TOut, Akka.NotUsed> OnFailuresWithBackoff<TIn, TOut, TMat>(System.Func<Akka.Streams.Dsl.Flow<TIn, TOut, TMat>> flowFactory, System.TimeSpan minBackoff, System.TimeSpan maxBackoff, double randomFactor) { }
+        public static Akka.Streams.Dsl.Flow<TIn, TOut, Akka.NotUsed> OnFailuresWithBackoff<TIn, TOut, TMat>(System.Func<Akka.Streams.Dsl.Flow<TIn, TOut, TMat>> flowFactory, System.TimeSpan minBackoff, System.TimeSpan maxBackoff, double randomFactor, int maxRestarts) { }
         public static Akka.Streams.Dsl.Flow<TIn, TOut, Akka.NotUsed> WithBackoff<TIn, TOut, TMat>(System.Func<Akka.Streams.Dsl.Flow<TIn, TOut, TMat>> flowFactory, System.TimeSpan minBackoff, System.TimeSpan maxBackoff, double randomFactor) { }
```
Comment (ismaelhamed) [inline src/core/Akka.API.Tests/CoreAPISpec.ApproveStreams.approved.txt]: The JVM doesn't have an overload. Should we just make a breaking change here in order to keep the API in sync?

**Comment by ismaelhamed** [comment]
> So I found out why TeamCity hates you - the dates on your commits are so old that TeamCity won't treat them like fresh targets for builds.

Is that a feature? :)

**Comment by Aaronontheweb** [comment]
> Is that a feature? :)

It just means that TeamCity is playing hard to get

File: src/core/Akka.API.Tests/CoreAPISpec.ApproveStreams.approved.txt (hunk 1 above)
Comment (Aaronontheweb) [reply src/core/Akka.API.Tests/CoreAPISpec.ApproveStreams.approved.txt]: @ismaelhamed IMHO, no - I think we should just add a new method. Better to not break the API if we can help it. Would you mind making that change?

Comment (ismaelhamed) [reply src/core/Akka.API.Tests/CoreAPISpec.ApproveStreams.approved.txt]: Done.

**Pull request merged**